File: pipeline/models/skills.py
"""Pydantic model for skills.json (includes characteristics)."""

from __future__ import annotations

from .common import CharacteristicGroup, CharacteristicId, DotRating, LenientModel, StrictModel


class Characteristic(LenientModel):
    """A characteristic (Strength, Dexterity, etc.)."""

    id: CharacteristicId
    name: str
    abbrev: str
    description: str
    specialties: list[str]
    ratings: list[DotRating]
    # Optional — only Intelligence has this
    notes: str | None = None


class Skill(StrictModel):
    """A skill entry."""

    id: str
    name: str
    characteristic: str  # CharacteristicId or "special"
    advanced: bool
    description: str
    specialties: list[str]


class SkillsFile(LenientModel):
    """Top-level structure of skills.json — contains both characteristics and skills."""

    characteristics: dict[CharacteristicGroup, list[Characteristic]]
    skills: dict[str, list[Skill]]
    skillNotes: dict | None = None

File: src/lib/dtd/schemas/skills.ts
/**
 * Zod schema for skills data — replaces former Pydantic model.
 */

import { z } from "zod";
import { CharacteristicGroup, CharacteristicId, DotRating } from "./common.ts";

// ---------------------------------------------------------------------------
// Characteristic (lenient)
// ---------------------------------------------------------------------------

export const Characteristic = z
	.object({
		id: CharacteristicId,
		name: z.string(),
		abbrev: z.string(),
		description: z.string(),
		specialties: z.array(z.string()),
		ratings: z.array(DotRating),
		notes: z.string().nullish(),
	})
	.passthrough();
export type Characteristic = z.infer<typeof Characteristic>;

// ---------------------------------------------------------------------------
// Skill (strict)
// ---------------------------------------------------------------------------

export const Skill = z
	.object({
		id: z.string(),
		name: z.string(),
		characteristic: z.string(),
		advanced: z.boolean(),
		description: z.string(),
		specialties: z.array(z.string()),
	})
	.strict();
export type Skill = z.infer<typeof Skill>;

// ---------------------------------------------------------------------------
// SkillsFile (lenient)
// ---------------------------------------------------------------------------

export const SkillsFile = z
	.object({
		characteristics: z.record(CharacteristicGroup, z.array(Characteristic)),
		skills: z.record(z.string(), z.array(Skill)),
		skillNotes: z.record(z.string(), z.unknown()).nullish(),
	})
	.passthrough();
export type SkillsFile = z.infer<typeof SkillsFile>;
